router/gossip.go

```diff
@@ -0,0 +1,149 @@
+package router
+
+import (
+	"bytes"
+	"encoding/gob"
+	"log"
+)
+
+type Gossip interface {
+	// specific message from one peer to another
+	// intermediate peers relay it using unicast topology.
+	GossipUnicast(dstPeerName PeerName, buf []byte) error
+	// send a message to every peer, relayed using broadcast topology.
+	GossipBroadcast(buf []byte) error
+}
+
+type Gossiper interface {
+	OnGossipUnicast(sender PeerName, msg []byte)
+	OnGossipBroadcast(msg []byte)
+	// Return state of everything we know; gets called periodically
+	Gossip() []byte
+	// merge in state and return "everything new I've just learnt",
+	// or nil if nothing in the received message was new
+	OnGossip(buf []byte) []byte
+}
+
+type GossipChannel struct {
+	ourself  *LocalPeer
+	name     string
+	hash     uint32
+	gossiper Gossiper
+}
+
+func (router *Router) NewGossip(channelName string, g Gossiper) Gossip {
+	h := hash(channelName)
+	channel := &GossipChannel{router.Ourself, channelName, h, g}
+	router.GossipChannels[h] = channel
+	return channel
+}
+
+func (router *Router) SendAllGossip() {
+	for _, c := range router.GossipChannels {
+		c.GossipMsg(c.gossiper.Gossip())
+	}
+}
+
+func (router *Router) SendGossip(channelName string, msg []byte) {
+	channelHash := hash(channelName)
+	if channel, found := router.GossipChannels[channelHash]; !found {
+		log.Println("[gossip] attempt to send on unknown channel:", channelName)
```

Comment:
I suggest introducing a function that can be called as `logGossip("attempt to send ...", channelName)`. There are potential call sites.